Abp.Dapper enhancement

osoykan opened the pull request:
* Dapper repository enhanced, some of methods deleted and renamed, this may be a breaking change to some of users

* Predicate generation enhanced

* Abp.Dapper **DynamicFilters** feature implemented, entities are automatically filtered now according to ISoftDelete, IMayHaveTenant, IMustHaveTenant, CreationAudit, ModificationAudit, DeletionAudit(thanks for inspiring @carldai0106 )

* The **DynamicFilters** are taken from CurrentUnitOfWork and shared with Abp.Dapper which means any filter change(disable, enable, set parameter) affects two ORM EF and Dapper both. 
 
* EntityChangeEventHelper integration implemented, entities are tracking now like as DbContext SaveChanges EntityTrackReport, Deleting-Deleted, Updating-Updated so on...

* Unit test added. This unit test is working on an **.mdf** file and contains single method to handle and prevent resouce breaks with multithreaded unit testing. (any better idea didn't come to my mind, any help would be appreciate this integration testing with dapper, but now it seems working.)

carldai0106 commented:
Hi @osoykan,

Wow, nice job. I like it. thanks for you refactoring my code.

carldai0106 commented:
Hi @hikalkan & @osoykan ,
Currently EF.DynamicFilter does not support EFCore, I think we can use this way (Dapper.DynamicFilters) to Abp.EFCore to filter some infomation.

hikalkan commented:
Thank you, I will review this when I have time.

hikalkan commented:
Thank you very much @osoykan for your awesome work :)

osoykan commented:
Thanks for merge @hikalkan. You're welcome, BTW new awesome features are coming on the way ! :) I've been waiting this merge to implement it. 

New roadmap is:

* Abstract out Abp.Dapper from EntityFramework
* Abp.Dapper should work with EF and NH both according the primary ORM choice and would be able to share their transactions.

We can talk and review this behaviour next PR :)

hikalkan commented:
Thank you, good news :)

hikalkan commented:
@osoykan can we simply use ``IsAssignableFrom`` instead of creating ``TypeExtensions.IsInheritsOrImplements`` extension method?

For example, instead of

``typeof(TEntity).IsInheritsOrImplements(typeof(ISoftDelete))``

can we use

``typeof(TEntity).IsAssignableFrom(typeof(ISoftDelete))``

Aren't they same?

osoykan commented:
@hikalkan,

Yes you're right, we can use. I've changed it and also fixed a problem that newly realized after my re-review.

PR is opened for these #2023.
